rfcs/webtransport_h3_test_server.md

```diff
@@ -0,0 +1,94 @@
+
+The server will be implemented under `tools/webtransport` directory.
+
+### Handlers
+
+A WebTransport handler is a Python script which contains a function `handle_event()`. Here is an example handler which echos back received data.
+
+```python
+from webtransport import (
+    WebTransportSession,
+    BidirectionalStreamDataReceived,
+    DatagramReceived,
+    WebTransportEvent
+)
+
+def handle_event(session: WebTransportSession,
+                 event: WebTransportEvent) -> None:
+    if isinstance(event, DatagramReceived):
+        session.send_datagram(data=event.data)
+    elif isinstance(event, BidirectionalStreamDataReceived):
+        session.send_stream_data(
+            stream_id=event.stream_id, data=event.data)
+```
+
+`session` is an object that represents a WebTransport over HTTP/3 session. It provides APIs to handle the session. Current proposed APIs can be found [here](https://bashi.github.io/wpt-wt-test-server-api/session.html).
+
+`event` represents a WebTransport event such as a reception of stream data. Current proposed events can be found [here](https://bashi.github.io/wpt-wt-test-server-api/events.html).
+
+#### Alternative handler design
+
+Alternatively, we can define a set of functions which are called upon each event. This API pattern is similar to existing [H2 handlers](https://web-platform-tests.org/writing-tests/h2tests.html). The equivalent of the above example might look like:
+
+```python
+def datagram_received(session: WebTransportSession,
+                      data: bytes):
+    session.send_datagram(data=data)
+
+
+def bidirectional_stream_data_received(session: WebTransportSession,
+                                       stream_id: int,
+                                       data: bytes):
+    session.send_stream_data(stream_id_id=stream_id, data=data)
+```
+
+A major downside of this approach is that keeping API documents and implementations in sync would require extra works, as H2 handler API document duplicates comments from the source.
+
+Another downside of this approach is that we won't have editor/IDE supports on what kind of callback functions are available.
+
+TODO: Spend more time on the API design. People may prefer this alternative approach.
+
+### Handler lookup
+
+When the server receives an [extended CONNECT method](https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3-01#section-3.2) for WebTransport, it looks up a corresponding handler from the directory specified by a configuration. The configuration is passed by `wptserve`. If the server finds a corresponding handler the server creates a new WebTransport session with the handler and starts passing events to the handler.
+
+The server uses `":path"` header field of an extended CONNECT request to look up a handler. For example, if `":path"` is `/echo`, the server looks up `echo.webtransport.py` from the handler directory.
```

Comment:
Why is this not just the wpt root, and the path contains the literal path to the Python file? e.g. if you want a handler in a file at `$WPT_ROOT/webtransport/handlers/echo.wt.py` you set the path to `/webtransport/handlers/echo.wt.py`? That seems easier for people debugging tests to understand (I think the websockets support acts more like you suggest, but that doesn't feel like a good design to copy since debugging those tests is always a pain for exactly this reason).

 Also note that your example here seems to be using a different naming convention to the section below.

Reply:
Makes sense. Updated to set the root directory as the same as wpt root.